### DISCUSSION TITLE
MacOS writeWithoutResponse

### DISCUSSION THREAD

**Discussion by JPHutchins** [hbldh/bleak#1589] [Q&A] [answered]
I am working on this issue: https://github.com/intercreate/smpclient/issues/30

Initiallly I suspected an issue wherein this send method:
```python
    async def send(self, data: bytes) -> None:
        for offset in range(0, len(data), self.mtu):
            await self._client.write_gatt_char(
                self._smp_characteristic, data[offset : offset + self.mtu], response=False
            )
```
was overwhelming the peripheral with the manual fragmentation getting sent as fast as possible at MTU.  After some experimentation I cannot reproduce the issue on the M3 that I'm borrowing, but obviously another user is able to reproduce reliably.

However, if that were the case, it appears that this callback should be called: https://developer.apple.com/documentation/corebluetooth/cbperipheraldelegate/peripheralisready(tosendwritewithoutresponse:)

So I added it to the PeripheralDelegate class:
```python
    def peripheralIsReady_toSendWriteWithoutResponse_(self, peripheral: CBPeripheral, characteristic: CBCharacteristic) -> None:
        logger.error("Called!")
```
But it is never called.  `didWriteValueForCharacterstic` is never called with an error, which I assume it would be if there was a problem with the rapid write without response.

I have also modified the Client class here:
```python
    async def write_gatt_char(
        self,
        characteristic: BleakGATTCharacteristic,
        data: Buffer,
        response: bool,
    ) -> None:
        value = NSData.alloc().initWithBytes_length_(data, len(data))
        await self._delegate.write_characteristic(
            characteristic.obj,
            value,
            (
                CBCharacteristicWriteWithResponse
                if response
                else CBCharacteristicWriteWithoutResponse
            ),
        )
        while self._peripheral.canSendWriteWithoutResponse() != True:
            logger.error("Not ready to send!")
            await asyncio.sleep(0.010)
        logger.debug(f"Write Characteristic {characteristic.uuid} : {data}")
```
That's using this API: https://developer.apple.com/documentation/corebluetooth/cbperipheral/cansendwritewithoutresponse

And interestingly I do see that the canSendWriteWithoutResponse flag is false for the first ~5 iterations after each write_gatt_char (~50ms).  I think it doesn't matter though because I suspect that the OS has some buffer and doesn't care that the write isn't going to be transmitted right away.  This would be more of an issue if you cared about some time sensitive write (in which case don't use a desktop OS and BLE).

Any help would be appreciated, thanks!
JP

**Comment by dlech** [comment, marked as answer]
As the name implies, when you write without a response, there is no way to get backpressure from the peripheral to know if it is keeping up with the incoming messages. So you can either do write with response and have it be slower or have some kind of occasional notification from the peripheral to indicate that it is ready for more data.

**Reply by JPHutchins** [reply]
Yes, it seems there is no error from MacOS or Bleak here.  No error is raised to Bleak and all of the transmissions work for me.  Primarily we've been testing Windows and Linux and hadn't sighted the problem before.  I suspect it may be that MacOS is "too good" at putting these fragmented packets "in the air".

@bjandre I noticed that your system was reporting an upload speed of ~24 KB/s which is way faster than mine was going at about ~13 KB/s.  Maybe that's why I'm not able to reliably reproduce.  Then again, I've seen Windows push 25 KB/s without issues.

At any rate, I find it curious that Zephyr isn't enabling write with response for this characteristic, given that it's for critical stuff like FW updates!  I haven't done much BLE work with Zephyr, but I think I have it correct that this source code is setting WRITE_WITHOUT_RESP and not WRITE_WITH_RESP: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/mgmt/mcumgr/transport/src/smp_bt.c#L356-L368

May be interesting to open a PR at Zephyr and add a Kconfig option that would enable `BT_GATT_CHRC_WRITE` on the characteristic.  Find def here: https://github.com/zephyrproject-rtos/zephyr/blob/fe03e0c4bb3f8a658efe69b04c430d1a15474c6c/include/zephyr/bluetooth/gatt.h#L295

Afterall, the client doesn't HAVE to use it, so I'm not sure of the downsides.

I think we have some good paths forward here!

**Reply by JPHutchins** [reply]
Opened a discussion over at Zephyr: https://github.com/zephyrproject-rtos/zephyr/discussions/73623

**Reply by dlech** [reply]
> suspect it may be that MacOS is "too good" at putting these fragmented packets "in the air".

Yeah, pretty sure we've seen users with something like this at Pybricks.

> At any rate, I find it curious that Zephyr isn't enabling write with response for this characteristic

The data rate would be painfully slow for something like transferring a large firmware file.

**Reply by JPHutchins** [reply]
> The data rate would be painfully slow for something like transferring a large firmware file.

How slow?  A "large" FW file may be about 500K.

In this protocol, every "request" receives a "response" (notify).  A lack of response, at least in this case, seems to indicate that MacOS overwhelmed the peripheral; the peripheral was not able to construct a complete request from the fragmented writes.  Typically the 2,475B "request" is fragmented across MTUs - ideally 5 495B MTUs.

Of course we can add retries to this specific routine - if a request times out we simply resend the entire request.  But it almost sounds like this particular issue could be prevented by only sending requests that fit in the MTU.  If a timeout did occur, MTU bytes would be retried rather than 2,475B.  May slow down the transfer somewhat, but so would retries.

It looks like Nordic's own iOS implementation uses a blocking delay on `canSendWriteWithoutResponse` 😮 https://github.com/NordicSemiconductor/IOS-nRF-Connect-Device-Manager/blob/30925fa37b453fafd34eb470d24851c6667d4848/Source/Bluetooth/McuMgrBleTransport.swift#L413-L435

FYI @bjandre - if Nordic's own implementation does this it may be worth looking into.  When I tried this it slowed the upload from ~13KB/s to ~5KB/s. We can work with Core Bluetooth without upstream changes to Bleak simply by accessing Core Bluetooth backend that Bleak provides.

**Reply by dlech** [reply]
> How slow?

You would have to measure it to see. It depends on a lot of factors.



> It looks like Nordic's own iOS implementation uses a blocking delay on `canSendWriteWithoutResponse`

Interesting. I would be interested to see Bluetooth packet captures comparing with this and without this.

Also, I wonder if we could use an observer to monitor the canSendWriteWithoutResponse` property instead of polling it similar to what we do for `isScanning` in the scanner since the delegate callback doesn't seem to do anything.

https://github.com/hbldh/bleak/blob/ce48c7fe470d7ead949d5249b74d571411defbb2/bleak/backends/corebluetooth/CentralManagerDelegate.py#L98
